app/server/lib/AttachmentFileManager.ts

```diff
@@ -135,10 +148,69 @@ export class AttachmentFileManager extends EventEmitter {
     fileExtension: string,
     fileData: Buffer
   ): Promise<AddFileResult> {
-    const fileIdent = await this._getFileIdentifier(fileExtension, Readable.from(fileData));
+    const fileIdent = await this._getFileIdentifier(fileExtension, stream.Readable.from(fileData));
     return this._addFile(storeId, fileIdent, fileData);
   }
 
+  /**
+   * Adds an attachment file to storage, if that attachment is known about but not currently
+   * available (as determined by `AttachmentFileManager.isFileAvailable`).
+   * This most frequently occurs when a document is (re)uploaded, meaning it has a new docPoolId
+   * and no copies of the attachment files.
+   *
+   * The file is restored to the store it was originally stored in, if that store is available.
+   * Otherwise, `defaultStoreId` is used.
+   *
+   * @param {string} fileIdent - Attachment file to attempt to restore.
+   * @param {stream.Readable} fileData - Contents of the file.
+   * @param {AttachmentStoreId | undefined} defaultStoreId - Store to use if the file's original store is unavailable.
+   * @returns {Promise<boolean>} - True if the file was added, false otherwise.
+   */
+  public async addMissingFileData(
+    fileIdent: string,
+    fileData: stream.Readable,
+    defaultStoreId: AttachmentStoreId | undefined,
+  ): Promise<boolean> {
+    const fileMetadata = await this._docStorage.getFileInfoNoData(fileIdent);
+    if (!fileMetadata) { return false; }
+    if (await this._isFileAvailable(fileMetadata)) { return false; }
+
+    const originalStoreExists = fileMetadata.ident in this._getStoreProvider().listAllStoreIds();
+    const newStoreId = originalStoreExists ? fileMetadata.ident : defaultStoreId;
+    const newStore = newStoreId && await this._getStore(newStoreId) || undefined;
+
+    if (newStoreId && !newStore) {
+      throw new StoreNotAvailableError(newStoreId);
+    }
+
+    // Internal storage is handled separately, as it needs the file as a Buffer in memory.
+    // External storage can avoid loading it into memory, as it's all stream APIs.
+    if (!newStore) {
```

Comment:
Can we invert this to put the positive branch of the if-else first? Getting rid of one negation makes this marginally easier to understand.

Reply:
Fine by me!